std/functional.d

```diff
@@ -1260,17 +1260,12 @@ Note:
 */
 template memoize(alias fun)
 {
-    import std.traits : ReturnType;
-     // https://issues.dlang.org/show_bug.cgi?id=13580
-    // alias Args = Parameters!fun;
-
-    ReturnType!fun memoize(Parameters!fun args)
+    auto memoize(Args...)(Args args) if (is(typeof(fun(args))))
     {
-        alias Args = Parameters!fun;
         import std.typecons : Tuple;
         import std.traits : Unqual;
 
-        static Unqual!(ReturnType!fun)[Tuple!Args] memo;
+        static Unqual!(typeof(fun(args)))[Tuple!Args] memo;
         auto t = Tuple!Args(args);
         if (auto p = t in memo)
             return *p;
```

Comment:
Lookup and optional insertion could be handled at once via `return memo.require(Tuple!Args(args), fun(args));`.

Reply:
Sadly, `require` isn't `nothrow`, so that's a `no-go`.